File: dynamicProgramming/climbingStairs.java
package dynamicProgramming;

// https://leetcode.com/problems/climbing-stairs/
public class climbingStairs {
    public static void main(String[] args) {

        SolutionCS solution  = new SolutionCS();
        System.out.println(solution.climbStairs(38));
    
    }
}

class SolutionCS {

    // Submitted by @Jiganesh


    //TC :O(2^n)
    //SC :O(n)
    // Using Recursion
    public int climbStairs(int n){
        return (n<=2) ? n : climbStairs(n-1) + climbStairs(n-2);
    }


    //TC :O(n)
    //SC :O(n)
    // Using DP and memoization
    public int climbStairsApproach2(int n) {
        
        if (n<=2) return n;
        int[] array = new int[n+1];
        array[0]=0;
        array[1]=1;
        array[2]=2;
        
        for (int i = 3 ; i<=n; i++){
            array[i]=array[i-1]+array[i-2];
        }
        
        return array[n]; 
            
    }
    
    // TC : O(N)
    // SC : O(1)

    // Modified Solution
    public int climbStairsApproach3(int n) {

        int a =1, b = 1;
        for (int i=2 ; i<=n; i++){
            int temp = a;
            a= b;
            b = temp +b;
        }
        return b;
        
    }
}

File: dynamicProgramming/climbingStairs.py
 
# https://leetcode.com/problems/climbing-stairs

class Solution(object):
    
    # Using recursion (brute Force)- Not efficient 
    # Will give TLE
    
    # TC : O(N!)
    # SC : O(N)
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        
        if n<=1 :
            return 1
        else :
            return self.climbStairs(n-1)+self.climbStairs(n-2)
        
    
    # TC :O(N)
    # SC : O(N)    
    def climbStairs (self,n):
        """
        :type n: int
        :rtype: int
        """
        dp = [1]*(n+1)
        for i in range(2, n+1):
            dp[i]= dp[i-1]+dp[i-2]
        return dp[n]
    
    #TC : O(N)
    #SC : O(1)
    
    # Runtime: 20 ms, faster than 56.28% of Python online submissions for Climbing Stairs.
    # Memory Usage: 13.2 MB, less than 88.62% of Python online submissions for Climbing Stairs.
    
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        a,b=1,1
        for _ in range(2, n+1):
            a, b = b, a+b
        return b

        
        
        
    
        
        
    